[JENKINS-35291] Initial API for BC 1.47

alvarolobato opened the pull request:
[JENKINS-35291](https://issues.jenkins-ci.org/browse/JENKINS-35291)

Dependency to multiple Bouncy Castle versions from jenkins core and plugins is causing problems due to the binary incompatibility between versions, the different supported algorithms, etc.

On Jenkins core 1.648, Bouncy Castle was bumped from version `1.47` to `1.54` as a result of a change in ``instance-identity`` module
* ``instance-identity-module`` bumped Bouncy Castle from `1.47` to `1.54`: [pom.xml (v 1.4)](https://github.com/jenkinsci/instance-identity-module/blob/instance-identity-1.4/pom.xml#L32) -> [pom.xml (v 1.5.1)](https://github.com/jenkinsci/instance-identity-module/blob/instance-identity-1.5.1/pom.xml#L33)
* Jenkins core bumped instance-identity-plugin:  [pom.xml (v 1.647) ](https://github.com/jenkinsci/jenkins/blob/stable-1.647/war/pom.xml#L107)-> [pom.xml (v 1.648)](https://github.com/jenkinsci/jenkins/blob/jenkins-1.648/war/pom.xml#L100)

**Problems found:**
* Plugins running in Jenkins < 1.648 get 1.47 from the parent class loader (in the default class loading strategy). If they include a later version in their own classpath, only "new" classes are seen from their referenced JAR.
* Plugins running in Jenkins >= 1.648 get 1.54 from the parent class loader (in the default class loading strategy). If they include an earlier version in their own classpath, only "removed" classes are seen from their referenced JAR.
* If the plugin is using JCA instead of BC classes directly usually the plugin works, but if ran in an BC 1.47 Jenkins core, there will be missing algorithms.
* Plugins are introducing different versions of BC which adds to de mix, potentially producing unpredictable results

**Some considerations:**
* Many plugins are only introducing BC dependency to do PEM encoding/decoding. This API got changed by the BC bump.
* instance-identity-module is using BC only for PEM encoding/decoding, removing this dependency from the core is possible

**Proposed solution**
A possible solution to this problem would to create a plugin (this plugin) from which plugins using BC will depend. Responsible for:
* Load BC into uber class loader.
* Register BC as a JVM security provider in order to allow other plugins to use JCA API with BC algorithms. 
* Provide an API to do common tasks like PEM Encoding/Decoding ensuring its stability among BC versions.

The implementation of this plugin does not expose any classes from BC, only JCA, allowing plugins to not depend on the specifics of BC.

Later some other actions would have to be taken in order to fully solve this situation, like having one version of the plugin for each version of BC, etc.

alvarolobato commented:
@reviewbybees

amuniz commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,349 @@
+    }
+    
+    /**
+     * Obtains raw JCA or BouncyCastle {@link Object} from the read PEM. 
```
Maybe a documented list of known return values is worth here, and a reminder that it depends on the nature of the PEM source.

alvarolobato replied:
@amuniz added a non exhaustive list of possible values

armfergom commented on pom.xml:
```diff
@@ -0,0 +1,96 @@
+    <dependency>
+      <groupId>org.bouncycastle</groupId>
+      <artifactId>bcpkix-jdk15on</artifactId>
+      <version>1.47</version>
```
Just a question. Given the following statements in the description:

> A possible solution to this problem would to create a plugin (this plugin) from which plugins using BC will depend

And 

> The implementation of this plugin **does not expose any classes from BC**, only JCA, allowing plugins to not depend on the specifics of BC.

Should not this be an `optional` dependency? If it is not `optional`, plugins depending on this plugin will have access to BC classes.

amuniz replied:
Right, that's the main point of this PR. Good catch.

alvarolobato replied:
Actually in this version it has to be runtime because the core is currently providing it.

amuniz replied:
Then it must be `<scope>provided</scope>`?

alvarolobato replied:
Yes, changing it.
El 6/6/2016 13:39, "Antonio" <notifications@github.com> escribió:

> In pom.xml
> <https://github.com/jenkinsci/bouncycastle-api/pull/1#discussion_r65875883>
> :
>
> > +      <url>https://repo.jenkins-ci.org/public/</url>
> > +    </repository>
> > +  </repositories>
> > +
> > +  <pluginRepositories>
> > +    <pluginRepository>
> > +      <id>repo.jenkins-ci.org</id>
> > +      <url>https://repo.jenkins-ci.org/public/</url>
> > +    </pluginRepository>
> > +  </pluginRepositories>
> > +
> > +  <dependencies>
> > +    <dependency>
> > +      <groupId>org.bouncycastle</groupId>
> > +      <artifactId>bcpkix-jdk15on</artifactId>
> > +      <version>1.47</version>
>
> Then it must be <scope>provided</scope>?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/jenkinsci/bouncycastle-api/pull/1/files/e3a0c8fb8b6a9b15c38033d17f9e17b91a893960#r65875883>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AQXNBBzc6zwfUKgDNtG2ZjBN4GjFlANqks5qJAbhgaJpZM4Iuv42>
> .
>

armfergom replied:
With either scopes (`<scope>compile</scope>` or `<scope>`provided</scope>`), the plugins using this would have direct access to BC classes which is one of the things to avoid, right? Could you please clarify? For me and others reading this review :)

alvarolobato replied:
On the current version of jenkins, Bouncy Castle is already provided by the core, so we have to keep it as provided.

Later when we remove it from the core, it will have to be compile scope as this plugin will be responsible for providing BC. If we put it optional we will no be sure that BC is available at runtime.

Plugins should not depend on the real BC clases, but the will be able to do it if they want, AFAIK there is no easy way to avoid this.

stephenc replied:
scope should be `provided` for now. when we move BC out of core this will have scope `compile`. Plugins can still depend on BC at their own risk. We need to give them a way to do that independent of Jenkins core as the issue that needs solving is that BC is bundled in core... not that BC is inherently evil (though it may well be - that is a question that is out of scope for this plugin to address ;-) )

amuniz commented:
LGTM, but delaying my bee until I see a use case in a downstream PR for some plugin using this API.

stephenc commented on src/main/java/jenkins/bouncycastle/api/BCPasswordException.java:
```diff
@@ -0,0 +1,45 @@
+
+package jenkins.bouncycastle.api;
+
+public class BCPasswordException extends RuntimeException {
```
javadoc. you are defining an API *everything* needs a spec in javadoc

alvarolobato replied:
Ups, adding it.

stephenc replied:
I believe this entire class can be deleted and its usage replaced with `java.security.UnrecoverableKeyException` or `java.security.UnrecoverableEntryException`

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,394 @@
+     * 
+     * @param pemFile {@link File} pointing to the PEM file to read
+     * @throws IOException launched if a problem exists reading the PEM information or the {@link File}
+     * @throws BCPasswordException in case PEM is passphrase protected and none or wrong is provided
```
`UnrecoverableKeyException` is a better exception to throw here. It matches exactly the problem and it is an existing JDK exception

stephenc replied:
So is the exception expected to be recoverable? I think so. If it is expected to be recoverable then it should be checked.

amuniz replied:
I think unchecked exceptions are better for public APIs in general, specially for maintainability and backward compatibility.

stephenc replied:
The critical test is can the caller be expected to recover from the exception.

If the caller cannot be expected to recover -> unchecked

If the caller can be expected to recover -> checked

If in doubt -> unchecked

In this case we are not in doubt. This is the same class of error as an IOException and we are assuming Java 7 for most users of this (as Java 7 is 1.612 or something like that) so multi-catch blocks should not be an issue.

Plus it is an existing exception that exactly matches the use case

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,394 @@
+     * @param pemFile {@link File} pointing to the PEM file to read
+     * @param passphrase passphrase for the encrypted PEM data. null if PEM data is not passphrase protected
+     * @throws IOException launched if a problem exists reading the PEM information or the {@link File}
+     * @throws BCPasswordException in case PEM is passphrase protected and none or wrong is provided
```
`UnrecoverableKeyException`

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,394 @@
+     * 
+     * @param pem {@link String} with the PEM data
+     * @throws IOException launched if a problem exists reading the PEM information
+     * @throws BCPasswordException in case PEM is passphrase protected and none or wrong is provided
```
`UnrecoverableKeyException`

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,394 @@
+     * @param pem {@link String} with the PEM data
+     * @param passphrase passphrase for the encrypted PEM data. null if PEM data is not passphrase protected
+     * @throws IOException launched if a problem exists reading the PEM information
+     * @throws BCPasswordException in case PEM is passphrase protected and none or wrong is provided
```
`UnrecoverableKeyException`

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,394 @@
+     * public keys.
+     */
+    @Nullable
+    public KeyPair getKeyPair() {
```
I would prefer something like `asKeyPair()` or `toKeyPair()`

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,394 @@
+ * Fingerprints, etc The supported algorithms will depend on the underlying version of BouncyCastle
+ *
+ */
+public class PEMManager {
```
This is not a `PEM Manager` it is closer to a `PEMEncodedObject` but it isn't one of them. Perhaps it is a `PEMEncodableObject`

stephenc replied:
On further consideration `PEMEncodable` may be a better name

alvarolobato replied:
Previously I had PEMCodec for encoder / decoder but didn't like it either.

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,394 @@
+    /**
+     * Stores the internal Bouncy Castle or JCA object
+     */
+    private Object pemObject;
```
just call it `object` because it is not PEM encoded in this form

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,394 @@
+/**
+ * A class that provides an API to manager PEM format, providing additional methods to handle Keys, Certificates,
+ * Fingerprints, etc The supported algorithms will depend on the underlying version of BouncyCastle
+ *
```
`@since` tag

stephenc commented on src/main/java/jenkins/bouncycastle/api/BCPasswordException.java:
```diff
@@ -24,6 +24,9 @@
+/**
+ * Indicates a password or passphrase is needed and none or a wrong one was provided
+ */
 public class BCPasswordException extends RuntimeException {
```
`java.security.UnrecoverableEntryException` or `java.security.UnrecoverableKeyException` is an existing exception that matches exactly the use case you had for this.

stephenc replied:
http://ted-gao.blogspot.ie/2013/03/best-practices-in-java-exception.html

> ### Don’t Create Custom Exception Unnecessarily
> If possible, try to use exceptions in stand JDK, well-established libraries or framework, rather than create new custom exception. When a new custom exception is created, it takes time for other people to learn it. It also increases the size of our source code. Doing so only when it brings true benefits that existing exception classes cannot bring.

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,394 @@
+     * @return {@link PublicKey} with the public key, null if a public key could not be obtained from the current data
+     */
+    @Nullable
+    public PublicKey getPublicKey() {
```
`asPublicKey()` or `toPublicKey()`

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,394 @@
+     * data
+     */
+    @Nullable
+    public Certificate getCertificate() {
```
`asCertificate()` or `toCertificate()`

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,394 @@
+     * data
+     */
+    @Nullable
+    public PrivateKey getPrivateKey() {
```
`asPrivateKey()` or `toPrivateKey()`

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,394 @@
+     * 
+     * @return {@link Object} read from the PEM
+     */
+    @Nullable
```
`asObject()` or `toObject()`

alvarolobato replied:
in this case we are not doing any conversion, is an actual get.

stephenc replied:
is this really `Nullable`? when is the constructor ever going to give a non-null object?

alvarolobato replied:
We rely on an external library that does the enconding and there are some parsers that in some cases return null. I can't ensure that it will be `Notnull`.

stephenc replied:
You can ensure in the constructor and throw an `IOException` or `UnrecoverableEntryException` if the parser gives you a `null`

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,394 @@
+     * @param data to be encoded
+     * @return hex formated string "ab:cd:ef:...:12"
+     */
+    public static String hexEncode(byte[] data) {
```
nullness annotations required here

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,394 @@
+     * @param data to be encoded
+     * @return base 64 formatted string
+     */
+    public static String encodeBase64(byte[] data) {
```
nullness annotations required here. also do we need to expose these as public methods?

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,394 @@
+     * @param data to be decoded
+     * @return decoded data
+     */
+    public static byte[] decodeBase64(String data) {
```
I do not think we need to expose this as a public method... in which case can we not just inline it

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,394 @@
+     * @throws IOException if a problem exists creating the digest
+     */
+    @Nonnull
+    public static byte[] getKeyDigest(@Nonnull Key k, @Nonnull String dg) throws IOException {
```
this one should throw the `NoSuchAlgorithmException` as the algorithm is user supplied (and call `dg` as `algorithm`)

alvarolobato replied:
Made my mind it was BC class don't know why... anyway, changed, pending commit.

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,394 @@
+     */
+    @Nonnull
+    public static byte[] getKeyDigestMD5(@Nonnull Key k) throws IOException {
+        return getKeyDigest(k, "MD5");
```
here you can catch the `NoSuchAlgorithmException` **because** the Java Language Specification states that MD5 must be supported

alvarolobato replied:
Changed, pending commit.

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,394 @@
+     */
+    @Nonnull
+    public static byte[] getKeyDigestSHA1(@Nonnull Key k) throws IOException {
+        return getKeyDigest(k, "SHA1");
```
here you can catch the `NoSuchAlgorithmException` **because** the Java Language Specification states that SHA1 must be supported

alvarolobato replied:
Changed, pending commit.

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,394 @@
+     * @return {@link PrivateKey} with the private key, null if a private key could not be obtained from the current
+     * data
+     */
+    @Nullable
```
`@CheckForNull`

stephenc replied:
this is `@CheckForNull` not `@Nullable` as the caller does not know before hand whether the PEM encoded object is a certificate or a key... thus *a priori* we do not know whether this could return `null` or `non-null` so the caller cannot be expected to know the null-ness and this means it cannot be `@Nullable` and must be `@CheckForNull`

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,394 @@
+     * @return {@link Certificate} with the certificate, null if a certificate could not be obtained from the current
+     * data
+     */
+    @Nullable
```
`@CheckForNull`

stephenc replied:
this is `@CheckForNull` not `@Nullable` as the caller does not know before hand whether the PEM encoded object is a certificate or a key... thus *a priori* we do not know whether this could return `null` or `non-null` so the caller cannot be expected to know the null-ness and this means it cannot be `@Nullable` and must be `@CheckForNull`

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,394 @@
+     * 
+     * @return {@link PublicKey} with the public key, null if a public key could not be obtained from the current data
+     */
+    @Nullable
```
this is `@CheckForNull` not `@Nullable` as the caller does not know before hand whether the PEM encoded object is a certificate or a key... thus *a priori* we do not know whether this could return `null` or `non-null` so the caller cannot be expected to know the null-ness and this means it cannot be `@Nullable` and must be `@CheckForNull`

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,394 @@
+     * @return {@link KeyPair} object with public and private keys or null if the read PEM didn't contain private and
+     * public keys.
+     */
+    @Nullable
```
this is `@CheckForNull` not `@Nullable` as the caller does not know before hand whether the PEM encoded object is a certificate or a key... thus *a priori* we do not know whether this could return `null` or `non-null` so the caller cannot be expected to know the null-ness and this means it cannot be `@Nullable` and must be `@CheckForNull`

alvarolobato replied:
changed.

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,394 @@
+     * @throws IOException launched if a problem exists generating the PEM information
+     */
+    @Nonnull
+    public String encodePEM() throws IOException {
```
the API feels unbalanced between these methods and the constructors. Perhaps just have a single constructor that takes the object and then make the `decode` methods static and returning the instance.

Also there is redundant use of PEM in the name... it's going to get really tiring typing

`PEMEncodableObject.decodePEM(string)`

Better would be `PEMEncodableObject.decode(string)`

Then we say... well `Object` is redundant so a better name for the class is `PEMEncodable`

alvarolobato replied:
Agree with the naming and sticking to just encode/decode changed.
About the constructor I did it this way because it is a way cleaner from an user view point which only wants to get a KeyPair or a PrivateKey for instance. Currently would be:

`new PEMManager(pemFile).toKeyPair()`

vs

`new PEMManager(PEMManager.decode(pemFile)).toKeyPair()`

or even
`Object object=PEMManager.decode(pemFile);
new PEMManager(object).toKeyPair()`

which would mean the user is using the object directly which we would like to avoid when possible.

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,394 @@
+     * 
+     * @throws IOException launched if a problem exists generating the PEM information or writing the {@link File}
+     */
+    public void encodePEM(@Nonnull File pemFile) throws IOException {
```
this is probably better being a method called `write` rather than `encode`

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,394 @@
+        this.pemObject = pemObject;
+    }
+
+    private void decodePEM(@Nonnull File pemFile, @Nullable String passphrase) throws IOException {
```
this should probably be `public static PEMEncodable read(File file, String passphrase) throws IOException, UnrecoverableEntryException`

stephenc commented on src/main/java/jenkins/bouncycastle/api/SecurityProviderInitializer.java:
```diff
@@ -0,0 +1,57 @@
+
+/**
+ * Initialization class to register Bouncy Castle as a security provider.
+ *
```
`@since`

amuniz replied:
Is `@since` necessary for a brand new APIs? I mean, everything without `@since` is intended to be there since the beginning of times.

stephenc replied:
always put `@since` on the class so people know when it was introduced

stephenc commented on src/main/java/jenkins/bouncycastle/api/SecurityProviderInitializer.java:
```diff
@@ -0,0 +1,57 @@
+     * 
+     */
+    @Initializer(before = InitMilestone.PLUGINS_STARTED)
+    public static void addSecurityProvider() {
```
`@Restricted(NoExternalUse.class)`

stephenc commented on src/main/java/jenkins/bouncycastle/api/SecurityProviderInitializer.java:
```diff
@@ -0,0 +1,57 @@
+ * Initialization class to register Bouncy Castle as a security provider.
+ *
+ */
+public class SecurityProviderInitializer extends Plugin {
```
We also need a `MasterToSlaveCallable` that will register BouncyCastle on the slave... perhaps that can be hidden and just expose a utility method that takes a `Channel` and will return when BouncyCastle is registered on the remote side... that way it can save a `ChannelProperty` on the local side of the channel to flag that BC has been registered so that a round-trip can be avoided for repeated calls and a lock can be managed in case multiple plugins try to register in parallel

stephenc commented on src/main/java/jenkins/bouncycastle/api/BCPasswordException.java:
```diff
@@ -0,0 +1,48 @@
+/**
+ * Indicates a password or passphrase is needed and none or a wrong one was provided
+ */
+public class BCPasswordException extends RuntimeException {
```
DELETE THIS CLASS

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,402 @@
+     * from the current PEM data.
+     * @throws IOException thrown if a problem exists creating the fingerprint
+     */
+    @Nullable
```
this is @CheckForNull not @Nullable as the caller does not know before hand whether the PEM encoded object is a certificate or a key... thus a priori we do not know whether this could return null or non-null so the caller cannot be expected to know the null-ness and this means it cannot be @Nullable and must be @CheckForNull

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,402 @@
+     * the current PEM data.
+     * @throws IOException if a problem exists creating the fingerprint
+     */
+    @Nullable
```
this is @CheckForNull not @Nullable as the caller does not know before hand whether the PEM encoded object is a certificate or a key... thus a priori we do not know whether this could return null or non-null so the caller cannot be expected to know the null-ness and this means it cannot be @Nullable and must be @CheckForNull

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,403 @@
+        try {
+            return getKeyDigest(k, "MD5");
+        } catch (NoSuchAlgorithmException e) {
+            throw new AssertionError("SHA1 algorithm not found to create digest");
```
This is MD5 and I would reference https://docs.oracle.com/javase/7/docs/api/java/security/MessageDigest.html where it says:

> Every implementation of the Java platform is required to support the following standard MessageDigest algorithms:
>
> * MD5
> * SHA-1
> * SHA-256

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,403 @@
+        try {
+            return getKeyDigest(k, "SHA1");
+        } catch (NoSuchAlgorithmException e) {
+           throw new AssertionError("SHA1 algorithm not found to create digest");
```
Reference the Javadoc for MessageDigest which says SHA-1 support is mandated

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,403 @@
+     * @return decoded data
+     */
+    @Nonnull
+    public static byte[] decodeBase64(@Nonnull String data) {
```
can be `private`

stephenc replied:
:bug:  do not expose more than we need to. Base64 encoding/decoding is not a responsibility of this API contract. By making these methods public we expose additional surface to the API and restrict the ability to refactor later

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,403 @@
+     * @return base 64 formatted string
+     */
+    @Nonnull
+    public static String encodeBase64(@Nonnull byte[] data) {
```
can be `private`

stephenc replied:
:bug: do not expose more than we need to. Base64 encoding/decoding is not a responsibility of this API contract. By making these methods public we expose additional surface to the API and restrict the ability to refactor later

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,403 @@
+    public static byte[] getKeyDigest(@Nonnull Key k, @Nonnull String algorithm) throws IOException, NoSuchAlgorithmException {
+        MessageDigest md5 = MessageDigest.getInstance(algorithm);
+
+        DigestInputStream in = new DigestInputStream(new ByteArrayInputStream(k.getEncoded()), md5);
```
inefficient.

```
MessageDigest md = MessageDigest.getInstance(algorithm);
md.update(k.getEncoded());
return md.digest();
```

then you have no intermediate garbage and you can loose the `IOException`

alvarolobato replied:
Good catch, didn't see that one. Changed.

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,403 @@
+    @Nonnull
+    public static String hexEncode(@Nonnull byte[] data) {
+        char[] hex = Hex.encodeHex(data);
+        StringBuilder buf = new StringBuilder();
```
we know the length: `hex.length + Math.max(0,hex.length / 2-1)`

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,403 @@
+        char[] hex = Hex.encodeHex(data);
+        StringBuilder buf = new StringBuilder();
+        for (int i = 0; i < hex.length; i += 2) {
+            if (buf.length() > 0)
```
`i > 0` is a better test as we do not need to invoke the function call

stephenc replied:
better still is to escape the first one from the loop and then the loop avoids the condition check... but if we make the test on `i > 0` then the JVM should make that optimization for us relatively quickly... so perhaps just leave the `i > 0` test in the loop

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,403 @@
+        StringBuilder buf = new StringBuilder();
+        for (int i = 0; i < hex.length; i += 2) {
+            if (buf.length() > 0)
+                buf.append(':');
```
braces should be used everywhere. do not try to save braces from `if` statements

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,400 @@
+     * @param k key to generate the digest from
+     * @param algorithm digest format
+     * @return the generated digest
+     * @throws IOException if a problem exists creating the digest
```
invalid javadoc

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,400 @@
+     * 
+     * @param k the key to generate the digest from
+     * @return the generated digest
+     * @throws IOException if a problem exists creating the digest
```
invalid javadoc

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,400 @@
+     * 
+     * @param k the key to generate the digest from
+     * @return the generated digest
+     * @throws IOException if a problem exists creating the digest
```
invalid javadoc

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,400 @@
+     * @param k the key to generate the digest from
+     * @return the generated digest
+     * @throws IOException if a problem exists creating the digest
+     * @see <a href="https://docs.oracle.com/javase/7/docs/api/java/security/MessageDigest.html"> JDK Digest
```
no. the exception thrown should have the link in the message. No caller of this should ever have to worry *because* the exception should never be thrown according to the Java Language Specification

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,400 @@
+     * @param k the key to generate the digest from
+     * @return the generated digest
+     * @throws IOException if a problem exists creating the digest
+     * @see <a href="https://docs.oracle.com/javase/7/docs/api/java/security/MessageDigest.html"> JDK Digest
```
Move this to the exception message

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,400 @@
+        try {
+            return getKeyDigest(k, "MD5");
+        } catch (NoSuchAlgorithmException e) {
+            throw new AssertionError("MD5 algorithm not found to create digest");
```
"MD5 algorithm support is mandated by Java Language Specification. See https://docs.oracle.com/javase/7/docs/api/java/security/MessageDigest.html"

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,400 @@
+     * 
+     * @return public key fingerprint in hex format "ab:cd:ef:...:12", null if the public key could not be obtained from
+     * the current PEM data.
+     * @throws IOException if a problem exists creating the fingerprint
```
invalid javadoc comment. we do not throw the exception any more

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,400 @@
+     * @throws IOException if a problem exists creating the fingerprint
+     */
+    @CheckForNull
+    public String getPublicKeyFingerprint() throws IOException {
```
redundant throws

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,400 @@
+     * @throws IOException thrown if a problem exists creating the fingerprint
+     */
+    @CheckForNull
+    public String getPrivateKeyFingerprint() throws IOException {
```
redundant throws

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,400 @@
+     * 
+     * @return {@link Object} read from the PEM
+     */
+    @CheckForNull
```
`@Nonnull`

stephenc replied:
:bug:

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,400 @@
+    /**
+     * Stores the internal Bouncy Castle or JCA object
+     */
+    private Object object;
```
@Nonnull

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,400 @@
+     * @throws IOException launched if a problem exists reading the PEM information or the {@link File}
+     * @throws BCPasswordException in case PEM is passphrase protected and none or wrong is provided
+     */
+    public PEMManager(@Nonnull File pemFile) throws IOException {
```
move to static method

alvarolobato replied:
About the constructor I did it this way because it is a way cleaner from an user view point which only wants to get a KeyPair or a PrivateKey for instance. Currently would be:

`new PEMManager(pemFile).toKeyPair()`

vs

`new PEMManager(PEMManager.decode(pemFile)).toKeyPair()`

or even

`Object object=PEMManager.decode(pemFile);
new PEMManager(object).toKeyPair()`

which would mean the user is using the object directly which we would like to avoid when possible.

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,400 @@
+     * @throws IOException launched if a problem exists reading the PEM information or the {@link File}
+     * @throws BCPasswordException in case PEM is passphrase protected and none or wrong is provided
+     */
+    public PEMManager(@Nonnull File pemFile, @Nullable String passphrase) throws IOException {
```
Move to static method

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,400 @@
+     * @throws IOException launched if a problem exists reading the PEM information
+     * @throws BCPasswordException in case PEM is passphrase protected and none or wrong is provided
+     */
+    public PEMManager(@Nonnull String pem) throws IOException {
```
Move to static method

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,400 @@
+     * @throws IOException launched if a problem exists reading the PEM information
+     * @throws BCPasswordException in case PEM is passphrase protected and none or wrong is provided
+     */
+    public PEMManager(@Nonnull String pem, @Nullable String passphrase) throws IOException {
```
Move to static method

alvarolobato commented:
@stephenc do we really want the user to use the API in this form?

`new PEMManager(PEMManager.decode(pemFile)).toKeyPair()``

or even

```
Object object=PEMManager.decode(pemFile);
new PEMManager(object).toKeyPair();
```

I personally think that this usage is much cleaner and intuitive:

`new PEMManager(pemFile).toKeyPair()`

amuniz commented:
> `new PEMManager(pemFile).toKeyPair()`

+1 to this API format. Much more intuitive.

@stephenc what's the good part of `new PEMManager(PEMManager.decode(pemFile)).toKeyPair()`?

reviewbybees commented:
You fe doing it wrong: `PEMEncodable.read(file).toKeyPair()`

Sent from my iPhone

> On 7 Jun 2016, at 11:59, Antonio <notifications@github.com> wrote:
> 
> new PEMManager(pemFile).toKeyPair()
> 
> +1 to this API format. Much more intuitive.
> 
> @stephenc what's the good part of new PEMManager(PEMManager.decode(pemFile)).toKeyPair()?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
> 
> -- 
> You received this message because you are subscribed to the Google Groups "engineering-code-reviews" group.
> To unsubscribe from this group and stop receiving emails from it, send an email to engineering-code-reviews+unsubscribe@cloudbees.com.
> To post to this group, send email to engineering-code-reviews@cloudbees.com.
> To view this discussion on the web visit https://groups.google.com/a/cloudbees.com/d/msgid/engineering-code-reviews/jenkinsci/bouncycastle-api/pull/1/c224247428%40github.com.

reviewbybees commented:
No `decode` should return a PEMEncodable and the File variant should be called `read`

Sent from my iPhone

> On 7 Jun 2016, at 11:40, Alvaro Lobato <notifications@github.com> wrote:
> 
> @stephenc do we really want the user to use the API in this form?
> 
> new PEMManager(PEMManager.decode(pemFile)).toKeyPair()`
> 
> or even
> 
> Object object=PEMManager.decode(pemFile);
> new PEMManager(object).toKeyPair();
> I personally think that this usage is much cleaner and intuitive:
> 
> new PEMManager(pemFile).toKeyPair()
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
> 
> -- 
> You received this message because you are subscribed to the Google Groups "engineering-code-reviews" group.
> To unsubscribe from this group and stop receiving emails from it, send an email to engineering-code-reviews+unsubscribe@cloudbees.com.
> To post to this group, send email to engineering-code-reviews@cloudbees.com.
> To view this discussion on the web visit https://groups.google.com/a/cloudbees.com/d/msgid/engineering-code-reviews/jenkinsci/bouncycastle-api/pull/1/c224243777%40github.com.

alvarolobato commented:
@stephenc this is more a mater of taste than any other thing. I'm going to change it and get this done.  

How should it be when the user wants to create a PEMEncodable with a KeyPair (ie), should he call the constructor or we override a decode(Object object) and make all the constructors privates?, in this case we are no decoding anything...

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,392 @@
+     *
+     * @param pemObject object to manage
+     */
+    public PEMManager(@Nonnull Object pemObject) {
```
Actually we should not even expose this constructor as public. Rather expose static factory methods: `PEMEncodable create(Key)`, `PEMEncodable create(KeyPair)`, `PEMEncodable create(Certificate)`. and leave this constructor private

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,392 @@
+ *
+ * @since 1.0
+ */
+public class PEMManager {
```
class should be `final`

stephenc replied:
I have repeatedly asked for this to be renamed to `PEMEncodable` I am :bug: on this current name

stephenc commented:
@alvarolobato 

`PEMEncodable.create(keyPair).write(file, passphrase)`

recena commented:
:bee: and 👍

amuniz commented:
:bee: (even 11 commits ago :smile:)

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMManager.java:
```diff
@@ -0,0 +1,392 @@
+     * @throws IOException launched if a problem exists reading the PEM information or the {@link File}
+     * @throws UnrecoverableKeyException in case PEM is passphrase protected and none or wrong is provided 
+     */
+    public PEMManager(@Nonnull File pemFile) throws IOException, UnrecoverableKeyException {
```
All constructors should be private or we will leak too much abstraction and restrict refactoring ability (e.g. if we start to add additional method signatures. constructors cannot have names. Use static factory methods like I have asked from the start) :bug:

stephenc commented:
:bug: in current state. to convert to a different vote:

1. Rename the class from `PEMManager` to `PEMEncodable` because that is what each instance holds. If it were a Manager then you would not create an instance for each encodable object. It is Encodable because the actual object it holds is not PEM encoded only capable of being encoded into that format
2. Remove the public constructors and make the class final - to narrow the API surface
3. Move to static factory methods: `PEMEncodable read(File)`, `PEMEncodable read(File,char[])`, `PEMEncodable decode(String)`, `PEMEncodable decode(String, char[])`, `PEMEncodable create(Certificate)`, `PEMEncodable create(Key)`, `PEMEncodable create(KeyPair)`
4. Remove the Base64 methods from the public API - we do not want 3rd parties using these for Base64 encoding/decoding as that is not part of our contract

stephenc commented:
w.r.t. naming things.

definition of manager is
> a person responsible for controlling or administering an organization or group of staff

if we apply that to PEM formats, a `PEMManager` would be responsible for controlling or administering a group of PEM encoded objects.

If the API is a `PEMManager` then that would be a static utility class that converts directly between the JCA objects and PEM encodings.

Representing each object individually is a better construct because it saves us from returning `Object` from the `read`/`decode` methods... but in doing that we no longer are dealing with a manager. instead we are dealing with an Object.... we could call it a `PEMObject` but what it holds is not a PEM object... rather it is something that can be encoded into PEM format. Thus the closest name we can get to is `PEMEncodable` but if you can come up with a better name I'm game to listen... this name is the local optimum that I have found and it is better than any alternatives I can find... thus in the absence of an even better name the correct name should be `PEMEncodable`

alvarolobato commented:
@stephenc I never said I wasn't going to change the name, in fact I've made the change a few hours ago, just hadn't committed it waiting for the rest of the review. I just didn't agree with some other changes. 

Shortly I will do a commit addressing all your suggestions, finishing all the javadocs and tests to make sure everything is ok.

alvarolobato commented:
@stephenc  I've addressed all your comments, let me know if I left anything behind.

Regarding the name it is hard to select one in fact I changed it multiple times during the development, as I told in a prior comment I had PEMCodec which is close to PEMEncoder, but ended changing because it has other connotations. Anyway, I've changed it to PEMEncodable.

stephenc commented on src/main/java/jenkins/bouncycastle/api/PEMEncodable.java:
```diff
@@ -0,0 +1,421 @@
+     * @throws UnrecoverableKeyException in case PEM is passphrase protected and none or wrong is provided
+     */
+    @Nonnull
+    public static PEMEncodable decode(@Nonnull String pem, @Nullable final String passphrase)
```
So this signature exposes a bad practice. When you store the passphrase in a `String` you have no control over the lifetime of the `String` in memory. Best practice is to pass the passphrase in a `char[]` and document in the Javadoc that the caller is responsible for zeroing out the `char[]` after use, e.g. with `Arrays.fill(passphrase, (char)0)`

By replicating this bad practice we do not enhance the security of consumers of this API

alvarolobato replied:
Changed to `char[]`

stephenc commented:
All good except for the password bad practice

stephenc commented:
:bee: 👍

amuniz commented:
:bee:
